File: tests/bindings/ruby/_coverage_runner.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'coverage'
require 'fileutils'

if ARGV.length != 2
  warn "usage: #{$PROGRAM_NAME} <test-script> <coverage-marshal>"
  exit 2
end

test_script = ARGV[0]
coverage_marshal = ARGV[1]

Coverage.start
status = 0

begin
  load test_script
rescue SystemExit => e
  status = e.status.nil? ? 1 : e.status.to_i
rescue Exception => e # rubocop:disable Lint/RescueException
  warn "#{e.class}: #{e.message}"
  warn e.backtrace.join("\n") if e.backtrace
  status = 1
ensure
  begin
    FileUtils.mkdir_p(File.dirname(coverage_marshal))
    File.binwrite(coverage_marshal, Marshal.dump(Coverage.result))
  rescue Exception => e # rubocop:disable Lint/RescueException
    warn "coverage dump failed: #{e.class}: #{e.message}"
    status = 1 if status.zero?
  end
end

exit status

File: tests/bindings/php/_coverage_runner.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

if ($argc !== 3) {
    fwrite(STDERR, "usage: {$argv[0]} <test-script> <coverage-json>\n");
    exit(2);
}

if (!function_exists('phpdbg_start_oplog') || !function_exists('phpdbg_end_oplog')) {
    fwrite(STDERR, "phpdbg oplog functions are unavailable\n");
    exit(2);
}

$testScript = $argv[1];
$coveragePath = $argv[2];
$exitStatus = 0;

phpdbg_start_oplog();

try {
    (static function (string $script): void {
        require $script;
    })($testScript);
} catch (Throwable $e) {
    $exitStatus = 1;
    fwrite(STDERR, get_class($e) . ': ' . $e->getMessage() . PHP_EOL);
    $trace = $e->getTraceAsString();
    if ($trace !== '') {
        fwrite(STDERR, $trace . PHP_EOL);
    }
}

$oplog = phpdbg_end_oplog();
if (!is_array($oplog)) {
    $oplog = [];
}

$coverageDir = dirname($coveragePath);
if (!is_dir($coverageDir) && !mkdir($coverageDir, 0777, true) && !is_dir($coverageDir)) {
    fwrite(STDERR, "failed to create coverage directory: {$coverageDir}\n");
    if ($exitStatus === 0) {
        $exitStatus = 1;
    }
}

$json = json_encode($oplog, JSON_PRETTY_PRINT);
if (!is_string($json) || file_put_contents($coveragePath, $json) === false) {
    fwrite(STDERR, "failed to write coverage file: {$coveragePath}\n");
    if ($exitStatus === 0) {
        $exitStatus = 1;
    }
}

exit($exitStatus);
